pygen/pygen_src/riscv_instr_stream.py

```diff
@@ -0,0 +1,230 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+Regression script for RISC-V random instruction generator
+"""
+
+from pygen_src.isa.riscv_instr import riscv_instr
+from pygen_src.riscv_instr_gen_config import riscv_instr_gen_config
+from pygen_src.riscv_instr_pkg import riscv_instr_name_t, riscv_instr_format_t,\
+    riscv_instr_category_t
+from riscv_instr_pkg import riscv_reg_t
+import random
+
+
+class riscv_instr_stream:
+    '''
+     Base class for RISC-V instruction stream
+     A instruction stream here is a  queue of RISC-V basic instructions.
+     This class also provides some functions to manipulate the instruction stream, like insert a new
+     instruction, mix two instruction streams etc.
+    '''
+
+    def __init__(self):
+        self.instr_list = []
+        self.instr_cnt = 0
+        self.label = ""
+        # User can specify a small group of available registers to generate various hazard condition
+        self.avail_regs = []
+        # Some additional reserved registers that should not be used as rd register
+        # by this instruction stream
+        self.reserved_rd = []
+        self.hart = 0
+
+    def initialize_instr_list(self, instr_cnt):
+        self.instr_list = []
+        self.instr_cnt = instr_cnt
+        self.create_instr_instance()
+
+    def create_instr_instance(self):
+        for i in range(self.instr_cnt):
+            instr = riscv_instr()
+            self.instr_list.append(instr)
+
+    def insert_instr(self, instr, idx=-1):
+        """
+           Insert an instruction to the existing instruction stream at the given index
+           When index is -1, the instruction is injected at a random location
+        """
+        current_instr_cnt = len(self.instr_list)
+        if idx == -1:
+            idx = random.randint(0, current_instr_cnt - 1)
+            while self.instr_list[idx].atomic:
+                idx = idx + 1
+                if idx == (current_instr_cnt - 1):
+                    self.instr_list.append(instr)
+                    return
+        elif idx > current_instr_cnt or idx < 0:
+            # TO DO: print an error
+            print("Error!")
+        self.instr_list.insert(idx, instr)
+
+    def insert_instr_stream(self, new_instr, idx=-1, replace=0):
+        """
+            Insert an instruction to the existing instruction stream at the given index
+            When index is -1, the instruction is injected at a random location
+            When replace is 1, the original instruction at the inserted position will be replaced
+        """
+        current_instr_cnt = len(self.instr_list)
+
+        if current_instr_cnt == 0:
+            self.instr_list = new_instr
+            return
+
+        if idx == -1:
+            idx = random.randint(0, current_instr_cnt - 1)
+            # cares must be taken to avoid targeting
```

Comment:
I think it will be better if we remove these empty extra in the code body. What do you think?

Reply:
We can remove also these lines. I have kept for better readability